rust_src/src/casetab.rs

```diff
@@ -3,100 +3,138 @@
     remacs_sys::EmacsInt,
     remacs_sys::{
-        map_char_table, set_char_table_extras, set_char_table_purpose, staticpro, Fcopy_sequence,
-        Fmake_char_table, Fset_char_table_range, CHAR_TABLE_SET,
+        map_char_table, staticpro, Fcopy_sequence, Fset_char_table_range, CHAR_TABLE_SET,
     },
     remacs_sys::{Qcase_table, Qcase_table_p, Qchar_table_extra_slots, Qnil},
+    symbols::LispSymbolRef,
     threads::ThreadState,
 };
 
+#[derive(Copy, Clone)]
 pub struct LispCaseTable(LispCharTableRef);
 
 impl LispCaseTable {
+    pub fn convert_char_table(mut table: LispCharTableRef) -> Self {
+        table.purpose = Qcase_table;
+        Self(table)
+    }
+
+    pub fn from_char_table(table: LispCharTableRef) -> Option<Self> {
+        if table.purpose.eq(Qcase_table) {
+            Some(Self(table))
+        } else {
+            None
+        }
+    }
+
+    pub fn is_proper(&self) -> bool {
+        let (up, canon, eqv) = self.extras();
+
+        (up.is_nil() || up.is_char_table())
+            && ((canon.is_nil() && eqv.is_nil())
+                || (canon.is_char_table() && (eqv.is_nil() || eqv.is_char_table())))
+    }
+
     pub fn extras(&self) -> (LispObject, LispObject, LispObject) {
         let extras = unsafe { self.0.extras.as_slice(3) };
         (extras[0], extras[1], extras[2])
     }
 
-    pub fn from_char_table(table: LispCharTableRef) -> Self {
-        Self(table)
+    pub fn set_extras(&mut self, idx: usize, value: LispObject) {
+        assert!(idx < 3);
+        let extras = unsafe { self.0.extras.as_mut_slice(3) };
+        extras[idx as usize] = value;
     }
 
     pub fn get(&self, idx: isize) -> LispObject {
         self.0.get(idx)
     }
-}
 
-impl From<LispObject> for LispCaseTable {
-    fn from(obj: LispObject) -> Self {
-        let case_table: Option<LispCharTableRef> = obj.into();
-        if !case_table_p(case_table) {
-            wrong_type!(Qcase_table_p, obj);
-        }
-        Self(case_table.unwrap())
-    }
-}
+    pub fn setup_extras(&mut self) {
+        let table = LispObject::from(*self);
 
-static mut Vascii_downcase_table: LispObject = Qnil;
-static mut Vascii_upcase_table: LispObject = Qnil;
-static mut Vascii_canon_table: LispObject = Qnil;
-static mut Vascii_eqv_table: LispObject = Qnil;
+        let (mut up, mut canon, mut eqv) = self.extras();
 
-fn set_case_table(table: LispObject, standard: bool) -> LispObject {
-    let case_table: LispCaseTable = table.into();
-    let (mut up, mut canon, mut eqv) = case_table.extras();
+        let sym_case_table = LispSymbolRef::from(Qcase_table);
 
-    unsafe {
         if up.is_nil() {
-            up = Fmake_char_table(Qcase_table, Qnil);
-            map_char_table(Some(set_identity), Qnil, table, up);
-            map_char_table(Some(shuffle), Qnil, table, up);
-            set_char_table_extras(table, 0, up);
+            let up_table = make_char_table(sym_case_table, Qnil);
+            up = LispObject::from(up_table);
```

Comment:
Is the type annotation here necessary? I think it could be removed in line 80 as well (`canon = LispObject::from(tem)`)